### ISSUE TITLE
Implementation of IUserService gets 403

### ISSUE THREAD

**Issue by scottblasingame** [IdentityServer/IdentityServer3#1551]
Got basics of a hybrid flow set up and working. Started to work in putting my custom user service that I need. I can login, but then redirect gets 403.

Here's the pertinent code from my User Service implementation (I'm just using a local list (Users) like in the samples just to get it running, so it's not making calls out yet):

        public Task AuthenticateLocalAsync(LocalAuthenticationContext context)
        {
            var user = Users.SingleOrDefault(x => x.Username == context.UserName && x.Password == context.Password);
            if (user == null)
            {
                return Task.FromResult<AuthenticateResult>(null);
            }
            else
            {
                context.AuthenticateResult = new AuthenticateResult(user.Subject, user.Username);
                return Task.FromResult(0);
            }

        }

        public Task GetProfileDataAsync(ProfileDataRequestContext context)
        {
            var user = Users.SingleOrDefault(x => x.Subject == context.Subject.GetSubjectId());
            if (user == null)
            {
                return Task.FromResult<IEnumerable<Claim>>(null);
            }

            context.IssuedClaims = user.Claims.Where(x => context.RequestedClaimTypes.Contains(x.Type));
            return Task.FromResult(0);
        }

        public Task IsActiveAsync(IsActiveContext context)
        {
            var user = Users.SingleOrDefault(x => x.Subject == context.Subject.GetSubjectId());
            context.IsActive = (user != null);
            return Task.FromResult(0);
        }

Here's my idsrv configuration:

        public void Configuration(IAppBuilder app)
        {
            var factory = InMemoryFactory.Create(
                clients: Clients.Get(),
                scopes: Scopes.Get());

            factory.UserService = new Registration<IUserService>(typeof(MyUserService));

            // Turn on logging
            LogProvider.SetCurrentLogProvider(new DiagnosticsTraceLogProvider());
            string logFilePath = AppDomain.CurrentDomain.BaseDirectory + "\\Logs\\log_idsrv.txt";
            FileStream hlogFile = new FileStream(logFilePath, FileMode.OpenOrCreate, FileAccess.Write);
            TextWriterTraceListener myListener = new TextWriterTraceListener(hlogFile);
            Trace.Listeners.Add(myListener);

            // Allow CORS
            factory.CorsPolicyService = new Registration<ICorsPolicyService>(new DefaultCorsPolicyService { AllowAll = true });
            
            var options = new IdentityServerOptions
            {
                SigningCertificate = Certificate.Load(),
                Factory = factory,
                
                LoggingOptions = new LoggingOptions
                {
                    EnableHttpLogging = true, 
                    EnableWebApiDiagnostics = true,
                    IncludeSensitiveDataInLogs = true
                },

            };
            
            app.UseIdentityServer(options);
        }

IIS gives this:

2015-06-29 21:34:16 ::1 GET /gr/CompileList - 443 Scott ::1 Mozilla/5.0+(Windows+NT+6.3;+WOW64;+Trident/7.0;+rv:11.0)+like+Gecko https://localhost/svc.security/connect/authorize?client_id=nedd_client&redirect_uri=https%3A%2F%2Flocalhost%2Fgr%2F&response_mode=form_post&response_type=code%20id_token%20token&scope=openid%20profile%20roles%20Web.Service.GroupRenewal&state=OpenIdConnect.AuthenticationProperties%3DDwjv9lMOtW2V7HVcO6lq82tCpV_t5I0F8DWuOQ-qM4ZnZ2Pz6BPSGpeBTJ0nin17nSlOXf3zrlb2xkevQQn5M-Vpp_C6qDcKR3ZKrHgO_XGxX1mbdqyqtoBdX-Maz8Kkcrqt3KTFvfjTTzhz82LVc_CNvB9P9kwbIFYMjBpZcEn_WFdSIcsoSjg0Ye7rLLCl8h-_cLSg0GOZC9JLZrJeyg&nonce=635712104437888887.ZTQwMmE5NzYtZDJkMS00MTU2LThjNTktMDBiYWQwMTY2N2ViNzIzOTNiNDUtNmJkZi00OTBmLTgwMGQtOWU3OTIxZjBjYTY5 403 0 0 14

Note I'm using dev channel (you'll see that I'm using the context vars). If I go back to the InMemory users, everything works fine. The IdSrv tracelog does not seem to indicate anything wrong, but I can send that along if needed.

Any idea what I'm doing wrong? Or how I should go about troubleshooting this? I've read through many articles and tried many things to no avail.

Thanks,
Scott

**Comment by brockallen** [comment]
Look at the 403's response headers and see what the error is.

**Comment by scottblasingame** [comment]
Using Fiddler just yields 403 Forbidden in the reponse headers. 

On the WebView I get this:

Detailed Error Information:

Module
   ManagedPipelineHandler 

Notification
   ExecuteRequestHandler 

Handler
   System.Web.Mvc.MvcHandler 

Error Code
   0x00000000 

Requested URL
   https://localhost:443/gr/CompileList 

Physical Path
   C:\Sandbox\f-GrRewrite\Solutions\DentalWeb\Presentation\Web.Presentation.GroupRenewal\CompileList 

Logon Method
   Cookies 

Logon User
   Scott 
--------------------------------------------
One thing that looks odd to me is that the cookies in the Request Header has about 8 entries for OpenIdConnect.nonce. Is that normal?

**Comment by scottblasingame** [comment]
Closing this. It was a really stupid error on my part as I did not bring over the Role claims properly to the UserService so my user was, in fact, not authorized. The only thing that bothers me was that I wasn't able to see the problem either in the IIS logs, in looking at the Request/Response Headers in Fiddler, or the trace logs for IdSrv. I even turned on IIS Failed Request Tracing, but got no help with that.
